pkg/maps/vtep/vtep.go

```
@@ -0,0 +1,134 @@
+// map value
+func (k Key) NewValue() bpf.MapValue { return &VtepEndpointInfo{} }
+
+func (k Key) String() string {
+	ipStr := net.IP(k.IP[:net.IPv4len]).String()
+	return fmt.Sprintf("%s", ipStr)
+}
+
+// NewKey returns an Key based on the provided IP address and mask.
+func NewKey(ip net.IP) Key {
+	result := Key{}
+
+	if ip4 := ip.To4(); ip4 != nil {
+		copy(result.IP[:], ip4)
+	}
+
+	return result
+}
+
+// VtepEndpointInfo implements the bpf.MapValue interface. It contains the
+// VTEP endpoint MAC and IP.
+// +k8s:deepcopy-gen=true
+// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapValue
+type VtepEndpointInfo struct {
+	VtepMAC        mac.Uint64MAC `align:"vtep_mac"`
+	TunnelEndpoint types.IPv4    `align:"tunnel_endpoint"`
+}
+
+func (v *VtepEndpointInfo) String() string {
+	return fmt.Sprintf("vtepmac=%s tunnelendpoint=%s",
+		v.VtepMAC, v.TunnelEndpoint)
+}
+
+// GetValuePtr returns the unsafe pointer to the BPF value.
+func (v *VtepEndpointInfo) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
+
+// Map represents an VTEP BPF map.
+type Map struct {
+	bpf.Map
+}
+
+// NewMap instantiates a Map.
+func NewMap(name string) *Map {
+	return &Map{
+		Map: *bpf.NewMap(
+			name,
+			bpf.MapTypeHash,
+			&Key{},
+			int(unsafe.Sizeof(Key{})),
+			&VtepEndpointInfo{},
+			int(unsafe.Sizeof(VtepEndpointInfo{})),
+			MaxEntries,
+			0, 0,
+			bpf.ConvertKeyValue,
+		).WithCache().WithPressureMetric().WithNonPersistent(),
+	}
+}
+
+var (
+	// VtepMAP is a mapping of all VTEP endpoint MAC, IPs
+	VtepMAP = NewMap(Name)
+)
+
+// Function to update vtep map with VTEP CIDR
+func UpdateVTEPMapping(newCIDR *cidr.CIDR, newTunnelEndpoint net.IP, vtepMAC mac.MAC) error {
+
+	key := NewKey(newCIDR.IP)
+
+	mac, err := vtepMAC.Uint64()
+	if err != nil {
+		return fmt.Errorf("invalid VTEP MAC: %v", err)
+	}
+
+	value := VtepEndpointInfo{
+		VtepMAC: mac,
+	}
+	if ip4 := newTunnelEndpoint.To4(); ip4 != nil {
+		copy(value.TunnelEndpoint[:], ip4)
+	}
```

Given that this function is checking for errors, and potentially returning errors if there are badly formatted values, if `ip4 == nil` here then it should return an error something like `fmt.Errorf("Cannot configure VTEP endpoint for IP %v", ip4")`.

---

since the `newTunnelEndpoint` passed on to `UpdateVTEPMapping` is already validated 
 to be valid IPv4 address in `pkg/option/config.go`, I am wondering if it is necessary to check `newTunnelEndpoint.To4()` error return here, thought?

---

Fair point, it shouldn't be necessary if the inputs are properly validated earlier. It's just odd to see that this function returns an error, but then there's an error case here that is not handled.

Though on closer inspection, it looks like this is the validation that occurs to check the addressing family in the option validation:

```
		if strings.Contains(ep, ":") {
			return fmt.Errorf("VTEP integration IPv6 not supported: %v", ep)
		}
```

I think that the option validation should do the same operation as you're doing here - after the `net.ParseIP()` call, it should attempt `endpoint.To4()` and check that it is non-nil. That way we can be sure that this code will always return a non-nil ipv4 address.